### ISSUE TITLE
OD Cost Matrix script failed (unknown reason)

### ISSUE THREAD

**Issue by no33mis** [Esri/public-transit-tools#170]
Hi, I am trying to perform the "Calculate Accessibility Matrix" and I always end up with the following error message: 

Validating OD Cost Matrix settings... 
Validating weight field Join_Count in destinations dataset... 
Inputs successfully validated. 
Copying origins to output... 
Precalculating network location fields for I:\...\temp.gdb\CalculateAccessibilityMatrix2... 
Precalculating network location fields for I:\...\scratch.gdb\TNAT_TempInput5... 
OD Cost Matrix script failed. 
Deleting intermediate outputs... 
Failed to execute (CalculateAccessibilityMatrix).

Since there is no obvious reason for the failure, I wonder if this is a known issue. 
Many thanks in advance.

**Comment by mmorang** [comment]
From this chain of messages, it seems like the failure is occurring in the parallelized OD cost matrix, but the messages aren't being passed back correctly. 

Can you confirm that you have the latest version of the toolbox (TransitNetworkAnalysisTools_0.7.1)?  I know I made a change back in April to improve exactly this issue, so I would expect better messaging.

You could also open up transit-network-analysis-tools\parallel_odcm.py in a text editor and verify that lines 1110-1114 look like this:

```python
    except Exception:  # pylint: disable=broad-except
        errs = traceback.format_exc().splitlines()
        for err in errs:
            LOGGER.error(err)
        raise
```

If you don't have the latest version or this code, try downloading the latest and see if that resolves the issue or at least improves the error message.

**Comment by no33mis** [comment]
Good morning, thank you for your suggestions. I can confirm that I have the latest version (0.7.1) and the lines within the parallel_odcm.py file are identical. I have tried to run the tool locally instead of a network drive, but got exactly the same error message. 

I have tried to run another tool now ("Calculate Travel Time Statistics") and interestingly it terminates with the same error message:

Validating OD Cost Matrix settings...
Inputs successfully validated.
Precalculating network location fields for I:\...\scratch.gdb\TNAT_TempInput...
Precalculating network location fields for I:\...\scratch.gdb\TNAT_TempInput0...
OD Cost Matrix script failed.
Deleting intermediate outputs...
Failed to execute (CalculateTravelTimeStatisticsOD).

I wonder if there is something not ok with my network (although the built in network analysis tools function perfectly). Are there any special settings I should be looking for?

**Comment by no33mis** [comment]
I have restarted the computer and at least now I am getting an error message, however only at a later stage:

Finished OD Cost Matrix calculation 121 of 122.
Finished OD Cost Matrix calculation 122 of 122.
Calculating statistics for final output...
Traceback (most recent call last):
  File "<string>", line 900, in execute
  File "C:\Temp\TransitNetworkAnalysisTools_0.7.1\CalculateODMatrixInParallel.py", line 438, in solve_large_od_cost_matrix
    self._execute_solve()
  File "C:\Temp\TransitNetworkAnalysisTools_0.7.1\CalculateODMatrixInParallel.py", line 613, in _execute_solve
    super()._execute_solve()
  File "C:\Temp\TransitNetworkAnalysisTools_0.7.1\CalculateODMatrixInParallel.py", line 388, in _execute_solve
    msg_string = output.strip().decode()
UnicodeDecodeError: 'utf-8' codec can't decode byte 0xe4 in position 81: invalid continuation byte
Failed to execute (CalculateAccessibilityMatrix).

The precalculation seems to be successful this time.

**Comment by mmorang** [comment]
Oh!  THIS problem.  Ugh, somebody else had this problem as well, and I was never able to reproduce it on my end, so I couldn't figure out how to fix it.

I don't know why you didn't see these details (the UnicodeDecodeError) the first time until you restarted your machine. My guess is some older version of the tool was cached.  Anyway, let's ignore that for now.

Here are the technical details of what's happening: The CalculateODMatrixInParallel.py file calls the parallel_odcm.py file as a subprocess.  This is a workaround to allow parallel_odcm.py to launch parallel processes, which you can't do from a script running directly in the Pro UI.  That subprocess logs various progress messages and passes them back to CalculateODMatrixInParallel.py, which writes them to the Pro geoprocessing pane's messages.  Something bad is happening when it tries to grab one of those logged messages because it can't decode it into the utf-8 character set.  So, one of the messages logged by the subprocess in parallel_odcm.py has some unrecognized character, and CalculateODMatrixInParallel.py chokes on it.

The real question is why parallel_odcm.py is writing the problematic message in the first place.  My guess is one of your filepaths or something in your data includes a weird character with a non-standard accent mark or umlaut or something.  The other person who had this problem eventually solved it by removing all non-ascii characters.

This SHOULD work.  I want to support all utf-8 characters, so I would really like to be able to reproduce the problem on my end.  I know you can't share your data, but if you're able to share any text strings with accent marks that you're using, that might help.

**Comment by no33mis** [comment]
Thanks once again for the detailed description. Really appreciate this support. I actually had the same thoughts about special characters (since we have some umlaut in the data). I have the created a new point layer with only the relevant information which are numeric, hence there are no strings anymore.

![image](https://user-images.githubusercontent.com/67404776/199772374-12538e3d-bca2-4682-8dec-031978238c56.png)

This is really anything that is left in the destination layer. The origin doesn't contain any special characters either. However, I still receive the same error message. 

I will once again shut down the computer tonight and will try again tomorrow with a different network dataset,

**Comment by mmorang** [comment]
My guess is the special character is in one of the filepaths, not in the data itself.  None of the messages the file generates should use the field values (I don't think...), but the filepaths are definitely used.

**Comment by no33mis** [comment]
I can confirm that this was the issue! There was a special character in the file path. I have copied the whole project and data into another local folder and tada, it worked. Many thanks once again for your help.

**Comment by mmorang** [comment]
Can you please tell me what the special character was?  I really want to fix this problem, but I can't do so until I can reproduce it reliably on my own machine.

Please paste the special character here in a GitHub comment.  Also, if you could paste a screenshot of what the character looks like on your machine, that would also help, since pasting it into GitHub might automatically alter it.

**Comment by mmorang** [comment]
I still can't reproduce the problem locally, but I've taken an educated guess about what's causing the problem and published a branch with an attempt at a fix.

If you have the time for a little bit of experimentation, could you check out this branch locally: https://github.com/Esri/public-transit-tools/tree/fix-encoding  Try re-running the tool using the project it its original location when it caused the problem and see if this fixes it.

**Comment by mmorang** [comment]
PS.  Before re-testing, make sure to close and reopen Pro or it won't pick up any of the code changes in the Python files.

**Comment by no33mis** [comment]
Hello, the special character was an "ä" - so a typical german umlaut. 
![image](https://user-images.githubusercontent.com/67404776/200494702-eb48b0f9-659a-4447-8201-49243da4dcd2.png)
Unfortunately, I again cannot share the complete file path due to security issues.

I will try to re-run the tool later this week with your modified version. I am super busy for the next few days with projects but will find some time for it on thu of fri.

**Comment by mmorang** [comment]
Another user was able to reproduce this problem and do some tests for me.  My latest change resolved it for her.  Please let me know if you experience it again and feel free to reopen this issue.
